### PR TITLE
MG-2441 - Add domain ID to API

### LINKED ISSUE #2441: Feature: Add Domain to API
### Is your feature request related to a problem? Please describe.

When using PAT this is the best approach to get Domain ID for authorization. Also this approach will be the most efficient for adding domains to events.

### Describe the feature you are requesting, as well as the possible use case(s) for it.

For example, when sharing a thing:
`http://localhost:9000/domains{{DOMAINID}}/things/{{THINGID}}/share`

### Indicate the importance of this feature to you.

Must-have

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by nyagamunene**
# What type of PR is this?
This is a feature because it adds the following functionality: it add domain id when passing requests in the API layer.

## What does this do?
It add domain id when passing requests in the API layer for users and things.

## Which issue(s) does this PR fix/relate to?

- Related Issue #2441 
- Resolves #2441 

## Have you included tests for your changes?
Yes.

## Did you document any new/modified feature?
Yes

### CODE REVIEW

**Comment by dborovcanin** [comment]
@nyagamunene Please make sure this is ready for review today. Also, please sign off your commits.

**Comment by nyagamunene** [comment]
> @nyagamunene Please make sure this is ready for review today. Also, please sign off your commits.

No problem

**Review by dborovcanin — CHANGES_REQUESTED** [review]
@nyagamunene Please rebase. This is to be merged before #2409.

File: users/api/clients.go [hunk 1]
```diff
@@ -85,6 +85,20 @@ func clientsHandler(svc users.Service, authn mgauthn.Authentication, tokenClient
...
+				api.EncodeResponse,
+				opts...,
+			), "list_clients").ServeHTTP)
+
+			r.Get("/search", otelhttp.NewHandler(kithttp.NewServer(
+				searchClientsEndpoint(svc),
+				decodeSearchClients,
+				api.EncodeResponse,
+				opts...,
+			), "search_clients").ServeHTTP)
```
Comment (dborovcanin) [inline users/api/clients.go]: Remove duplicate handlers.

**Review by dborovcanin — CHANGES_REQUESTED** [review]
Overall, looks good. Fix CI and comments and it's ready for another review. I'll also check if there is a better way to handle things in Bootstrap.

**Review by arvindh123 — CHANGES_REQUESTED** [review]
For Things, groups ,channels  
DomainID should be automatically decoded from URL , 
We can do that in Middleware and add to authn.Session

**Comment by nyagamunene** [comment]
> For Things, groups ,channels DomainID should be automatically decoded from URL , We can do that in Middleware and add to authn.Session

Since in this PR we are just adding the domainID to URL should I make this change in the other PR?

File: users/api/requests.go [hunk 2]
```diff
@@ -89,9 +89,13 @@ type listMembersByObjectReq struct {
 	mgclients.Page
 	objectKind string
 	objectID   string
+	domainID   string
 }
 
 func (req listMembersByObjectReq) validate() error {
+	if req.domainID == "" {
```
Comment (arvindh123) [inline users/api/requests.go]: @nyagamunene  We have changes in  request in this PR , It is not relevant to this PR, please remove it

**Review by arvindh123 — CHANGES_REQUESTED** [review]

**Comment by dborovcanin** [comment]
> > For Things, groups ,channels DomainID should be automatically decoded from URL , We can do that in Middleware and add to authn.Session
> 
> Since in this PR we are just adding the domainID to URL should I make this change in the other PR?

Since we are already breaking API, you can do it here. In the next one, you can remove the domain from the token, just like planned.

File: cmd/things/main.go [hunk 3]
```diff
@@ -231,7 +231,6 @@ func main() {
 	g.Go(func() error {
 		return server.StopSignalHandler(ctx, cancel, logger, svcName, httpSvc)
 	})
```
Comment (arvindh123) [inline cmd/things/main.go]: Please undo this empty changes

File: consumers/notifiers/api/requests.go [hunk 4]
```diff
@@ -51,5 +51,6 @@ func (req listSubsReq) validate() error {
 	if req.token == "" {
 		return apiutil.ErrBearerToken
 	}
+
```
Comment (arvindh123) [inline consumers/notifiers/api/requests.go]: Please remove this empty changes , Is this due to linter?

File: pkg/authn/authn.go [hunk 5]
```diff
@@ -14,6 +14,11 @@ type Session struct {
 	SuperAdmin   bool
 }
 
+func (s *Session) UpdateSession(DomainID string) {
```
Comment (arvindh123) [inline pkg/authn/authn.go]: No need of this changes function, 

Instead check and invalidate authn session to empty object , if authn session domain id is not equal to domainID 

We could not have these logic for all endpoints, it is needed only required for domains, things, groups and channels 

For users this logic is no needed logic.

File: certs/api/requests.go [hunk 6]
```diff
@@ -39,22 +44,30 @@ func (req addCertsReq) validate() error {
...
+	thingID  string
+	domainID string
+	pm       certs.PageMetadata
 }
 
 func (req *listReq) validate() error {
 	if req.pm.Limit > maxLimitSize {
 		return apiutil.ErrLimitSize
 	}
+	if req.domainID == "" {
```
Comment (arvindh123) [inline certs/api/requests.go]: Certs service doesn't have authn middleware ? Do we need token here?

File: bootstrap/api/requests.go [hunk 7]
```diff
@@ -28,6 +29,10 @@ func (req addReq) validate() error {
 		return apiutil.ErrBearerToken
 	}
 
+	if req.domainID == "" {
```
Comment (arvindh123) [inline bootstrap/api/requests.go]: Bootstrap service doesn't have authn middleware ? Do we need token here?

**Review by arvindh123 — CHANGES_REQUESTED** [review]

File: certs/api/requests.go (hunk 6 above)
Comment (nyagamunene) [reply certs/api/requests.go]: > Certs service doesn't have authn middleware ? Do we need token here?

DomainID is needed for issuing and revoking cert

File: internal/groups/api/requests.go [hunk 8]
```diff
@@ -160,5 +160,6 @@ func (req listMembersReq) validate() error {
 	if req.groupID == "" {
 		return apiutil.ErrMissingID
 	}
+
```
Comment (arvindh123) [inline internal/groups/api/requests.go]: remove this line

**Review by arvindh123 — COMMENTED** [review]
LGTM

File: users/api/clients.go [hunk 9]
```diff
@@ -164,7 +171,7 @@ func clientsHandler(svc users.Service, authn mgauthn.Authentication, tokenClient
 		// SpiceDB provides list of user ids in given user_group_id
 		// and users service can access spiceDB and get the user list with user_group_id.
 		// Request to get list of users present in the user_group_id {groupID}
-		r.Get("/groups/{groupID}/users", otelhttp.NewHandler(kithttp.NewServer(
+		r.Get("/domains/{domainID}/groups/{groupID}/users", otelhttp.NewHandler(kithttp.NewServer(
```
Comment (dborovcanin) [inline users/api/clients.go]: There is no need for `/domains` prefix here, use `/{domainID}/groups/...`. I.e. unlike other entities, let's not prefix domain ID.

Comment (dborovcanin) [reply users/api/clients.go]: Same for other endpoints.

**Review by dborovcanin — CHANGES_REQUESTED** [review]

**Review by dborovcanin — APPROVED** [review]

**Pull request merged**